From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Sun, 28 Aug 2022 11:36:28 +0200
Subject: [PATCH 1/1] Add `_emscripten_get_dynamic_libraries_js` helper

Handy for determining which dynamic libraries are loaded in native code.

Upstream-Status: Pending

diff --git a/src/library_dylink.js b/src/library_dylink.js
index 1111111..2222222 100644
--- a/src/library_dylink.js
+++ b/src/library_dylink.js
@@ -896,6 +896,19 @@ var LibraryDylink = {
     });
   },
 
+  _emscripten_get_dynamic_libraries_js__deps: ['$allocateUTF8'],
+  _emscripten_get_dynamic_libraries_js__sig: 'i',
+  _emscripten_get_dynamic_libraries_js: function() {
+    var size = (dynamicLibraries.length + 1) * {{{ Runtime.POINTER_SIZE }}};
+    var libs = {{{ makeMalloc('_emscripten_get_dynamic_libraries_js', 'size') }}};
+    var libs_ptr = libs >> {{{ POINTER_SHIFT }}};
+    dynamicLibraries.forEach((lib) => {
+      {{{ POINTER_HEAP }}}[libs_ptr++] = {{{ to64('allocateUTF8(lib)') }}};
+    });
+    {{{ POINTER_HEAP }}}[libs_ptr] = {{{ to64('0') }}};
+    return libs;
+  },
+
   // void* dlopen(const char* filename, int flags);
   $dlopenInternal__deps: ['$FS', '$ENV', '$dlSetError', '$PATH'],
   $dlopenInternal: function(handle, jsflags) {
